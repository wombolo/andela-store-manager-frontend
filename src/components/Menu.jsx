import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

export class Menu extends Component {
  handleLogout = () => {
    localStorage.clear();
    window.location.reload();
  };

  headerLinks = ({isAuthenticated, user}) => {
    let links;

    if (!isAuthenticated) {
      links = (
        <React.Fragment>
          <ul>
            <li><NavLink to="/" className='links'>Home</NavLink></li>
            <li><NavLink to="/login" className='links'>Login</NavLink></li>
          </ul>
        </React.Fragment>);
    } else {

      const {profile} = user;
      const AdminLinks = [
        { to: '/add-product', name: 'Add New Product'},
        { to: '/add-profile', name: 'Add New User'},
      ];

      const AttendantLinks =[
        // { to: '/all-sales', name: 'My Sales'}
      ];

      const thisUserLinks = profile.role === 'admin' ? AdminLinks : AttendantLinks;

      links = (
        <React.Fragment>
          <ul>
            <li><NavLink to="/" className='links'>Home</NavLink></li>
            <li><NavLink to="/dashboard" className='links'>Dashboard</NavLink></li>
          </ul>

          <ul className='links'>
            <li><NavLink to="/all-products" className='links'>All Products</NavLink></li>
            {thisUserLinks.map(link => (
              <li key={link.to}><NavLink to={link.to} className='links'> {link.name} </NavLink></li>
            ))}

            <li><NavLink to={'/all-sales'} className='links'> All Sales </NavLink></li>
            <li><NavLink to="/login" className='links' onClick={this.handleLogout}>Logout</NavLink></li>
          </ul>
        </React.Fragment>);
    }
    return links;
  };

  render() {
    return (
      <div className='sidebar-menu'>
        {this.headerLinks(this.props.auth)}
      </div>
    )
  }
}


Menu.propTypes = {
  auth: PropTypes.shape({}).isRequired,
};

export const mapStateToProps = state => ({
  auth: state.auth,
});

export default connect(mapStateToProps)(Menu);
