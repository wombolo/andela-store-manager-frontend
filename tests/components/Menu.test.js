import 'babel-polyfill'
import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import { MemoryRouter } from 'react-router-dom';

import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';

import { Menu } from '../../src/components/Menu.jsx';

let wrapper;

const props = {
  auth: {
    user: {
      profile: {
        id: 1,
        firstname: 'Akano',
        lastname: 'Peter',
        email: 'elkenzom1@gmail.com',
        role: 'admin',
        image: 'team-7.jpg',
        password: null
      },
      iat: 1551809044,
      exp: 1551895444
    },
    isAuthenticated: true,
    error: '',
    isLoading: false
  },
};


const mockStore = configureStore([thunk]);

const store = mockStore({ ...props });


describe('Menu Component', () => {
  it('should match snapshot', () => {
    wrapper = shallow(<Menu {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Methods in Class', () => {
    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter>
          <Menu {...props} />
          </MemoryRouter>
        </Provider>,
      );
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it('triggers handleLogout()', () => {
      const MenuComp = wrapper.find('Menu');
      sinon.spy(MenuComp.instance(), 'handleLogout');

      MenuComp.instance().handleLogout();
      expect(MenuComp.instance().handleLogout.calledOnce)
        .toEqual(true);
    });

    //
    // it('triggers handleRemoveFromMenu()', () => {
    //   const MenuComp = wrapper.find('Menu');
    //
    //   sinon.spy(MenuComp.instance(), 'handleRemoveFromMenu');
    //
    //   MenuComp.instance().handleRemoveFromMenu(1);
    //   expect(MenuComp.instance().handleRemoveFromMenu.calledOnce)
    //     .toEqual(true);
    //
    //   // const button = MenuComp.find('button').last();
    //   // button.simulate('click', 1);
    // });
    //
    //
    // it('triggers handleSubmit()', () => {
    //   const MenuComp = wrapper.find('Menu');
    //
    //   sinon.spy(MenuComp.instance(), 'handleSubmit');
    //   MenuComp.instance().handleSubmit({preventDefault: () => {}});
    //   expect(MenuComp.instance().handleSubmit.calledOnce)
    //     .toEqual(true);
    // });
  });


  describe('Methods in Class', () => {

    beforeEach(() => {
      props.auth.isAuthenticated=false;
      props.auth.user.profile.role='store_attendant';

      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter>
          <Menu {...props} />
          </MemoryRouter>
        </Provider>,
      );
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it('show links for unauthenticated user', () => {
      const MenuComp = wrapper.find('Menu');
      sinon.spy(MenuComp.instance(), 'handleLogout');

      MenuComp.instance().handleLogout();
      expect(MenuComp.instance().handleLogout.calledOnce)
        .toEqual(true);
    });
  });
});
