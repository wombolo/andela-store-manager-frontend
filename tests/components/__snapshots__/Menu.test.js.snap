// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component should match snapshot 1`] = `
<div
  className="sidebar-menu"
>
  <ul>
    <li>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="links"
        to="/"
      >
        Home
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="links"
        to="/dashboard"
      >
        Dashboard
      </NavLink>
    </li>
  </ul>
  <ul
    className="links"
  >
    <li>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="links"
        to="/all-products"
      >
        All Products
      </NavLink>
    </li>
    <li
      key="/add-product"
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="links"
        to="/add-product"
      >
         
        Add New Product
         
      </NavLink>
    </li>
    <li
      key="/add-profile"
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="links"
        to="/add-profile"
      >
         
        Add New User
         
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="links"
        to="/all-sales"
      >
         All Sales 
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="links"
        onClick={[Function]}
        to="/login"
      >
        Logout
      </NavLink>
    </li>
  </ul>
</div>
`;
